In [71]:
# import necessary libraries 
import pandas as pd 
import numpy as np 

In [72]:
# load and read the data
df = pd.read_excel("collection_orders_test.xlsx")

In [73]:
# explore the data
df.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
0        269210  242624 2022-03-10 22:01:11 2022-03-10 22:14:30       0   
1        253601  184788 2022-02-20 21:16:03 2022-02-21 16:40:05       0   
2        250270  174631 2022-02-16 02:13:03 2022-02-16 14:22:03       0   
3        251559  178209 2022-02-18 19:32:04 2022-02-18 22:10:47       0   
4        248466  173047 2022-02-13 23:19:03 2022-02-13 23:25:07       0   

   initial_amount  final_amount  retailer_id               status  agent_id  \
0          1000.0         500.0       118272  PARTIALLY_COLLECTED      2231   
1          1000.0           0.0        37800               FAILED      1950   
2          1000.0        1000.0        37800            COLLECTED      2288   
3          1000.0        1000.0        37800            COLLECTED      2288   
4          1000.0         500.0        37800  PARTIALLY_COLLECTED       869   

   district_id  
0         87.0  
1         35.0  
2         35.0  
3         35.0  
4         35.0

In [74]:
df.describe()

run_sheet_id          co_id         delays  initial_amount  \
count  131693.000000  131693.000000  131693.000000   131693.000000   
mean   256440.071234  206012.728695       0.067111     1362.468291   
std     34459.469064  123350.661014       0.250214      697.937277   
min     63129.000000      19.000000       0.000000        0.000000   
25%    229221.000000  118625.000000       0.000000     1000.000000   
50%    261039.000000  220574.000000       0.000000     1000.000000   
75%    286277.000000  270625.000000       0.000000     1500.000000   
max    308757.000000  487993.000000       1.000000    10000.000000   

        final_amount    retailer_id       agent_id    district_id  
count  131693.000000  131693.000000  131693.000000  130136.000000  
mean     1300.645176   96839.840888    1981.358698     127.545775  
std       958.895600   88307.792350     613.493759     162.091945  
min         0.000000      67.000000      72.000000       0.000000  
25%      1000.000000   24517.000000    1645.000000      31.000000  
50%      1000.000000   60220.000000    2068.000000      55.000000  
75%      1500.000000  175698.000000    2447.000000     106.000000  
max     30000.000000  295843.000000    3078.000000     732.000000

In [75]:
# Check the data types 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131693 entries, 0 to 131692
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   run_sheet_id    131693 non-null  int64         
 1   co_id           131693 non-null  int64         
 2   created_at      131693 non-null  datetime64[ns]
 3   updated_at      131693 non-null  datetime64[ns]
 4   delays          131693 non-null  int64         
 5   initial_amount  131693 non-null  float64       
 6   final_amount    131693 non-null  float64       
 7   retailer_id     131693 non-null  int64         
 8   status          131693 non-null  object        
 9   agent_id        131693 non-null  int64         
 10  district_id     130136 non-null  float64       
dtypes: datetime64[ns](2), float64(3), int64(5), object(1)
memory usage: 11.1+ MB


In [76]:
df.shape

(131693, 11)

In [77]:
# Checking for nulls 
df.isnull().sum()

run_sheet_id         0
co_id                0
created_at           0
updated_at           0
delays               0
initial_amount       0
final_amount         0
retailer_id          0
status               0
agent_id             0
district_id       1557
dtype: int64

we can notice here we only have null valeus in the " district_id " column, we won't drop them as we dont want to lose any data so we will just replace them with "unknow"


In [78]:
df[df["district_id"].isnull()]

run_sheet_id   co_id          created_at          updated_at  delays  \
684           249512  174277 2022-02-15 17:01:03 2022-02-15 17:10:27       0   
1255          296094  360985 2022-04-05 14:25:05 2022-04-05 14:31:19       0   
1256          299583  396867 2022-04-08 19:15:04 2022-04-09 00:26:12       0   
1257          305118  445101 2022-04-13 18:28:03 2022-04-14 00:44:25       0   
1350          304786  444548 2022-04-13 17:20:06 2022-04-13 17:33:28       0   
...              ...     ...                 ...                 ...     ...   
131322        236737  150011 2022-01-27 14:56:03 2022-01-27 15:30:57       0   
131323        234185  147564 2022-01-23 16:45:03 2022-01-23 19:04:28       0   
131324        292792  309412 2022-04-01 21:09:03 2022-04-02 13:44:17       0   
131479        290362  286989 2022-03-30 19:54:04 2022-03-30 20:53:54       0   
131480        278471  256237 2022-03-20 18:12:10 2022-03-20 18:19:05       0   

        initial_amount  final_amount  retailer_id               status  \
684             1000.0        1000.0       214855            COLLECTED   
1255            1000.0         500.0       287576  PARTIALLY_COLLECTED   
1256            1000.0         500.0       287576  PARTIALLY_COLLECTED   
1257            1000.0         500.0       287576  PARTIALLY_COLLECTED   
1350            1000.0         500.0       288441  PARTIALLY_COLLECTED   
...                ...           ...          ...                  ...   
131322          1000.0        1000.0       238208            COLLECTED   
131323          1000.0        1000.0       238208            COLLECTED   
131324          1000.0        1000.0       238208            COLLECTED   
131479          1000.0         500.0       238895  PARTIALLY_COLLECTED   
131480          1000.0        1000.0       238895            COLLECTED   

        agent_id  district_id  
684         2396          NaN  
1255        2140          NaN  
1256        2068          NaN  
1257        2068          NaN  
1350        2874          NaN  
...          ...          ...  
131322      2007          NaN  
131323      2007          NaN  
131324      2675          NaN  
131479      2880          NaN  
131480      2204          NaN  

[1557 rows x 11 columns]

In [79]:
# fill the null values with "unknown"
df.fillna("unknown", inplace = True)

In [80]:
# checking
df.isna().any()

run_sheet_id      False
co_id             False
created_at        False
updated_at        False
delays            False
initial_amount    False
final_amount      False
retailer_id       False
status            False
agent_id          False
district_id       False
dtype: bool

In [81]:
# Checking for duplicates
df.duplicated().any()

False

In [82]:
# explore delays column
df["delays"].value_counts()

0    122855
1      8838
Name: delays, dtype: int64

In [83]:
# explore status column
df["status"].value_counts()

COLLECTED              112992
PARTIALLY_COLLECTED      9352
FAILED                   9349
Name: status, dtype: int64

in this case we will consider "COLLECTED" and "PARTIALLY_COLLECTED" is a success and other is a fail
we will replace "COLLECTED " and "PARTIALLY_COLLECTED " with 1 and "FAILED" with 0 

In [84]:
# make a copy of the Data 
df_copy = df.replace({"status" :{"COLLECTED" : 1,"PARTIALLY_COLLECTED" : 1 ,"FAILED" :0 }})

In [85]:
df_copy.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
0        269210  242624 2022-03-10 22:01:11 2022-03-10 22:14:30       0   
1        253601  184788 2022-02-20 21:16:03 2022-02-21 16:40:05       0   
2        250270  174631 2022-02-16 02:13:03 2022-02-16 14:22:03       0   
3        251559  178209 2022-02-18 19:32:04 2022-02-18 22:10:47       0   
4        248466  173047 2022-02-13 23:19:03 2022-02-13 23:25:07       0   

   initial_amount  final_amount  retailer_id  status  agent_id district_id  
0          1000.0         500.0       118272       1      2231          87  
1          1000.0           0.0        37800       0      1950          35  
2          1000.0        1000.0        37800       1      2288          35  
3          1000.0        1000.0        37800       1      2288          35  
4          1000.0         500.0        37800       1       869          35

In [86]:
# figuring out the success rate and failure rate 
df_copy["status"].value_counts(normalize = True)

1    0.929009
0    0.070991
Name: status, dtype: float64

##### So as we can see here the success rate is 92.3 % while the failure or error rate is 07.09 % 
and we will make further analyis to figure out what causes this error rate


In [87]:
# Checking number of unique agents 
df["agent_id"].nunique()

312

In [88]:
# Checking number of unique retailers 
df["retailer_id"].nunique()

11662

i assume the updated_at column refers to when the collector actually showed up at the retailer and started to collect, so we will add a new column which is the difference between updated at and created at to see how fast the the collector respond to the request 

In [89]:
# add the new column "response_in_hours"
df["response_in_hours"] = df["updated_at"] - df["created_at"]

In [90]:
df.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
0        269210  242624 2022-03-10 22:01:11 2022-03-10 22:14:30       0   
1        253601  184788 2022-02-20 21:16:03 2022-02-21 16:40:05       0   
2        250270  174631 2022-02-16 02:13:03 2022-02-16 14:22:03       0   
3        251559  178209 2022-02-18 19:32:04 2022-02-18 22:10:47       0   
4        248466  173047 2022-02-13 23:19:03 2022-02-13 23:25:07       0   

   initial_amount  final_amount  retailer_id               status  agent_id  \
0          1000.0         500.0       118272  PARTIALLY_COLLECTED      2231   
1          1000.0           0.0        37800               FAILED      1950   
2          1000.0        1000.0        37800            COLLECTED      2288   
3          1000.0        1000.0        37800            COLLECTED      2288   
4          1000.0         500.0        37800  PARTIALLY_COLLECTED       869   

  district_id response_in_hours  
0          87          00:13:19  
1          35          19:24:02  
2          35          12:09:00  
3          35          02:38:43  
4          35          00:06:04

In [91]:
# we will get the average hours for response 
df["response_in_hours"].mean()

Timedelta('0 days 14:57:44.875999')

as we can see it has almost 15 hours in average 

In [92]:
# get only the data where status = FAILED
df.query("status == 'FAILED'")

run_sheet_id   co_id          created_at          updated_at  delays  \
1             253601  184788 2022-02-20 21:16:03 2022-02-21 16:40:05       0   
8             242372  162422 2022-02-04 15:02:03 2022-02-05 18:40:47       0   
66            267845  235452 2022-03-08 16:51:03 2022-03-09 12:32:46       0   
72            266973  228039 2022-03-06 15:27:09 2022-03-08 15:22:25       1   
97            222426   96188 2022-01-03 11:26:03 2022-01-03 19:24:59       0   
...              ...     ...                 ...                 ...     ...   
131564        293998  331597 2022-04-03 16:18:04 2022-04-03 22:44:14       0   
131600        260084  217551 2022-03-01 13:52:03 2022-03-01 15:23:51       0   
131618        296307  333433 2022-04-03 20:02:03 2022-04-05 23:42:34       1   
131626        289077  274370 2022-03-28 15:53:03 2022-03-29 22:22:21       1   
131691        252875  181259 2022-02-19 22:01:03 2022-02-20 20:04:45       0   

        initial_amount  final_amount  retailer_id  status  agent_id  \
1               1000.0           0.0        37800  FAILED      1950   
8               1000.0           0.0       244267  FAILED      2181   
66              1500.0           0.0         7357  FAILED      2506   
72              1500.0           0.0         7357  FAILED       391   
97              1000.0           0.0        60203  FAILED      2015   
...                ...           ...          ...     ...       ...   
131564          1000.0           0.0       273630  FAILED      2184   
131600          1000.0           0.0       248076  FAILED      1877   
131618          2000.0           0.0       274131  FAILED      2941   
131626          1500.0           0.0       248511  FAILED      2331   
131691          1000.0           0.0       249112  FAILED      2287   

       district_id response_in_hours  
1               35   0 days 19:24:02  
8               60   1 days 03:38:44  
66              38   0 days 19:41:43  
72              38   1 days 23:55:16  
97              55   0 days 07:58:56  
...            ...               ...  
131564         555   0 days 06:26:10  
131600          61   0 days 01:31:48  
131618         551   2 days 03:40:31  
131626          31   1 days 06:29:18  
131691         105   0 days 22:03:42  

[9349 rows x 12 columns]

In [93]:
# Save the new data in df_failed
df_failed = df.query("status == 'FAILED'")

In [94]:
# get only the data where status = "collecter or partially collected"
df_success = df.query("status == 'COLLECTED'" or "status == 'PARTIALLY_COLLECTED'")

In [95]:
df_success.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
2        250270  174631 2022-02-16 02:13:03 2022-02-16 14:22:03       0   
3        251559  178209 2022-02-18 19:32:04 2022-02-18 22:10:47       0   
5        268848  238415 2022-03-09 11:30:10 2022-03-10 17:55:05       1   
6        286288  268778 2022-03-26 10:08:03 2022-03-27 17:37:56       0   
7        279318  255246 2022-03-20 00:59:03 2022-03-21 18:17:30       0   

   initial_amount  final_amount  retailer_id     status  agent_id district_id  \
2          1000.0        1000.0        37800  COLLECTED      2288          35   
3          1000.0        1000.0        37800  COLLECTED      2288          35   
5          1000.0        1000.0       244267  COLLECTED      2483          60   
6          1000.0        1000.0       244267  COLLECTED      2820          60   
7          1000.0        1000.0       244267  COLLECTED      2327          60   

  response_in_hours  
2   0 days 12:09:00  
3   0 days 02:38:43  
5   1 days 06:24:55  
6   1 days 07:29:53  
7   1 days 17:18:27

In [96]:
print(df_failed["response_in_hours"].mean())
print(df_success["response_in_hours"].mean())
print(df["response_in_hours"].mean())

0 days 21:29:58.574286
0 days 14:44:36.509248
0 days 14:57:44.875999


as it was expected the data with failed status has more average in time response comparing to the original data and the data with success status, which means that time for response is really important in order to make it a success status

In [97]:
# getting the number of delays in the data with failed status
df_failed["delays"].value_counts()

0    7232
1    2117
Name: delays, dtype: int64

In [98]:
# the number of delays overall
df["delays"].value_counts()

0    122855
1      8838
Name: delays, dtype: int64

##### We can notice here thats almost 24 % of the total delays come with failed status, which means when the order doesn’t have a terminal status from the first attempt there is a chance of 24 % that this order will fail


In [99]:
# getting data when the delay = 1, which means when the order didn't get a terminal status from the first attempt
df.query("delays == 1")

run_sheet_id   co_id          created_at          updated_at  delays  \
5             268848  238415 2022-03-09 11:30:10 2022-03-10 17:55:05       1   
13            287355  270714 2022-03-26 21:13:04 2022-03-28 19:16:36       1   
18            286574  271062 2022-03-27 00:07:03 2022-03-27 19:55:23       1   
40            256532  193442 2022-02-23 13:32:04 2022-02-25 15:44:15       1   
52            237503  149088 2022-01-26 00:01:12 2022-01-29 14:10:08       1   
...              ...     ...                 ...                 ...     ...   
131625        274192  247067 2022-03-14 01:45:10 2022-03-16 12:38:19       1   
131626        289077  274370 2022-03-28 15:53:03 2022-03-29 22:22:21       1   
131629        268844  235705 2022-03-08 18:04:03 2022-03-10 13:59:47       1   
131689        273390  246946 2022-03-13 23:51:03 2022-03-15 19:49:25       1   
131690        254552  186190 2022-02-21 07:28:12 2022-02-22 18:41:44       1   

        initial_amount  final_amount  retailer_id               status  \
5               1000.0        1000.0       244267            COLLECTED   
13              1000.0         500.0        48165  PARTIALLY_COLLECTED   
18              1000.0        1000.0       189156            COLLECTED   
40              2000.0        2000.0         5470            COLLECTED   
52              2000.0        2000.0         5470            COLLECTED   
...                ...           ...          ...                  ...   
131625          1000.0        1000.0       248511            COLLECTED   
131626          1500.0           0.0       248511               FAILED   
131629          1000.0        1000.0       248511            COLLECTED   
131689          1000.0        1000.0       249112            COLLECTED   
131690          1000.0        1000.0       249112            COLLECTED   

        agent_id district_id response_in_hours  
5           2483          60   1 days 06:24:55  
13          2891          46   1 days 22:03:32  
18          2602          87   0 days 19:48:20  
40          1654          31   2 days 02:12:11  
52          1654          31   3 days 14:08:56  
...          ...         ...               ...  
131625      2482          31   2 days 10:53:09  
131626      2331          31   1 days 06:29:18  
131629      2232          31   1 days 19:55:44  
131689      2139         105   1 days 19:58:22  
131690      2331         105   1 days 11:13:32  

[8838 rows x 12 columns]

In [100]:
# save the new data in df_delays
df_delays = df.query("delays ==1")

In [101]:
df_delays.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
5         268848  238415 2022-03-09 11:30:10 2022-03-10 17:55:05       1   
13        287355  270714 2022-03-26 21:13:04 2022-03-28 19:16:36       1   
18        286574  271062 2022-03-27 00:07:03 2022-03-27 19:55:23       1   
40        256532  193442 2022-02-23 13:32:04 2022-02-25 15:44:15       1   
52        237503  149088 2022-01-26 00:01:12 2022-01-29 14:10:08       1   

    initial_amount  final_amount  retailer_id               status  agent_id  \
5           1000.0        1000.0       244267            COLLECTED      2483   
13          1000.0         500.0        48165  PARTIALLY_COLLECTED      2891   
18          1000.0        1000.0       189156            COLLECTED      2602   
40          2000.0        2000.0         5470            COLLECTED      1654   
52          2000.0        2000.0         5470            COLLECTED      1654   

   district_id response_in_hours  
5           60   1 days 06:24:55  
13          46   1 days 22:03:32  
18          87   0 days 19:48:20  
40          31   2 days 02:12:11  
52          31   3 days 14:08:56

In [102]:
# getting more info about the status in the data where delays = 1
df_delays["status"].value_counts()

COLLECTED              5873
FAILED                 2117
PARTIALLY_COLLECTED     848
Name: status, dtype: int64

In [103]:
# getting more info about the status in the original data 
df["status"].value_counts()

COLLECTED              112992
PARTIALLY_COLLECTED      9352
FAILED                   9349
Name: status, dtype: int64

##### here we can say that 22.6% of the total fails happend when there was a delay 

In [104]:
# save df_failed data in csv file to load it in Power Bi and do further analysis
df_failed.to_csv("df_failed.csv")

we will make further analysis regarding the agents who have an id number 493 and 1737 as it showed that 211 of the failed orders were happened when he was the one responsible for it and 208 failed orders for the agent 1737, as they are the highest two agents with failed orders 

In [105]:
# get only the data where agent_id = 493
df_agent_493 = df.query("agent_id == 493")

In [106]:
df_agent_493.head()

run_sheet_id  co_id          created_at          updated_at  delays  \
41         221526  94760 2021-12-31 20:18:03 2022-01-01 18:11:06       0   
347        201492   9970 2021-11-21 20:09:03 2021-11-21 20:19:10       0   
611        188853   3926 2021-10-26 17:17:09 2021-10-26 17:26:30       0   
623        188827   3891 2021-10-26 14:20:05 2021-10-26 14:28:38       0   
894        213989  22436 2021-12-16 20:48:03 2021-12-16 20:54:53       0   

     initial_amount  final_amount  retailer_id     status  agent_id  \
41           2000.0        2000.0         5470  COLLECTED       493   
347          1000.0        1000.0        30617  COLLECTED       493   
611          1000.0        1000.0        21098  COLLECTED       493   
623          1000.0        1000.0        16119  COLLECTED       493   
894          1000.0        1000.0        81922  COLLECTED       493   

    district_id response_in_hours  
41           31          21:53:03  
347          62          00:10:07  
611          62          00:09:21  
623          62          00:08:33  
894          25          00:06:50

In [107]:
# finding the average time response for this agent
df_agent_493["response_in_hours"].mean()

Timedelta('0 days 23:15:58.864705')

as we can see the mean of the response time for this agent is very high which totally affect the status of the order which most likely to be failed

In [108]:
df_agent_493["delays"].value_counts()

0    679
1      1
Name: delays, dtype: int64

so now we made sure the reason behind the failed orders with this agent is probably the late time response, cuz here it shows that he always complete the order in the first attempt 

In [109]:
df_agent_493["status"].value_counts()

COLLECTED              425
FAILED                 211
PARTIALLY_COLLECTED     44
Name: status, dtype: int64

from this info we can say that almost 31 % of the total orders that this agent was responsible for led to a failure

-------------------------------------------------------------------------------------------------------------------------------

##### for the agent 1737

In [110]:
# get only the data where agent_id = 1737
df_agent_1737 = df.query("agent_id == 1737")
df_agent_1737.head()


run_sheet_id   co_id          created_at          updated_at  delays  \
203        223729   97386 2022-01-05 12:51:03 2022-01-05 17:13:49       0   
218        247074  171420 2022-02-11 17:19:03 2022-02-12 14:01:59       0   
225        232658  145817 2022-01-20 13:49:03 2022-01-20 16:09:58       0   
227        230030  131378 2022-01-16 14:06:12 2022-01-16 18:27:27       0   
257        220731   92323 2021-12-29 17:13:04 2021-12-29 19:28:46       0   

     initial_amount  final_amount  retailer_id     status  agent_id  \
203          2000.0        2000.0         5233  COLLECTED      1737   
218          2000.0        2200.0         5233  COLLECTED      1737   
225          1500.0        1500.0         5233  COLLECTED      1737   
227          1500.0        1500.0         5233  COLLECTED      1737   
257          1000.0        1000.0        12943  COLLECTED      1737   

    district_id response_in_hours  
203          75          04:22:46  
218          75          20:42:56  
225          75          02:20:55  
227          75          04:21:15  
257          37          02:15:42

In [111]:
df_agent_1737["response_in_hours"].mean()

Timedelta('0 days 13:22:36.753095')

the average time response for this agent is not that high comparing to the agent 493

In [112]:
df_agent_1737["delays"].value_counts()

0    2501
1      83
Name: delays, dtype: int64

for this agent the number of delays is relatively high comparing to the previous agent, with a 3.3 % of the total orders he was responsible for 

In [113]:
# we will get the data of this agent where delays = 1
df_1737_delay = df_agent_1737.query("delays ==1")

In [114]:
# we will get number of retailers that this agent had delays with
df_1737_delay["retailer_id"].nunique()

71

In [115]:
# we will get the retailer that this agent had the most delay with
df_1737_delay["retailer_id"].value_counts()

30925     4
188379    3
47196     2
85816     2
164780    2
         ..
28876     1
24004     1
30146     1
223613    1
8576      1
Name: retailer_id, Length: 71, dtype: int64

the retailer with id number 30925 is the one that this agent had the most delays count with 

In [116]:
df_agent_1737["status"].value_counts()

COLLECTED              2323
FAILED                  208
PARTIALLY_COLLECTED      53
Name: status, dtype: int64

from this we can say that 8 % of the total orders that this agent was responsible for led to a failure 

-------------------------------------------------------------------------------------------------------------------------------

we will make other further analysis but this time regarding the districts id number 54 and 28 as it showed that 447 failed orders  that occurred in the district 54, and 391 failed orders in the district 28 as they are the highest two districts with failed orders

In [117]:
# get only the data where district_id = 54
df_district_54 = df.query("district_id == 54")

In [118]:
df_district_54.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
483        270253  240356 2022-03-10 01:43:03 2022-03-12 14:35:11       1   
685        307547  472252 2022-04-15 22:20:06 2022-04-16 12:43:52       0   
686        299199  367282 2022-04-06 00:15:03 2022-04-08 23:15:37       1   
687        295175  349320 2022-04-04 21:01:03 2022-04-04 21:11:51       0   
688        304888  433177 2022-04-12 16:54:03 2022-04-13 22:47:36       1   

     initial_amount  final_amount  retailer_id     status  agent_id  \
483          1000.0           0.0        32419     FAILED      2017   
685         10000.0       16350.0        51535  COLLECTED      2290   
686          7000.0        8000.0        51535  COLLECTED      2759   
687          1000.0        1000.0        51535  COLLECTED      2724   
688          9000.0        9200.0        51535  COLLECTED      1521   

    district_id response_in_hours  
483          54   2 days 12:52:08  
685          54   0 days 14:23:46  
686          54   2 days 23:00:34  
687          54   0 days 00:10:48  
688          54   1 days 05:53:33

In [119]:
df_district_54["response_in_hours"].mean()

Timedelta('0 days 15:03:39.734829')

we can the mean time for response is not actually that high

In [120]:
df_district_54["status"].value_counts(normalize = True)

COLLECTED              0.816708
FAILED                 0.092893
PARTIALLY_COLLECTED    0.090399
Name: status, dtype: float64

the percentage of the failed is not big comparing the the total 

In [121]:
# find out numbers of retailers that located in this district
df_district_54["retailer_id"].nunique()

389

we will get the district data where status = failed to find out number or retailers that the agents failed to complete the order in with in this district


In [122]:
df_54_failed = df_district_54.query("status == 'FAILED'")

In [123]:
# we will find out the number of retailers 
df_54_failed["retailer_id"].nunique()

194

so almost 50 % of the retailers in the district id number 54 have a failed status which means the agents failed to complete the order with them

-------------------------------------------------------------------------------------------------------------------------------

##### for the district 28

In [124]:
# get only the data where district_id = 28
df_district_28 = df.query("district_id == 28")
df_district_28.head()


run_sheet_id   co_id          created_at          updated_at  delays  \
114        283485  264604 2022-03-24 18:57:04 2022-03-24 20:19:11       0   
235        298142  379434 2022-04-06 23:25:04 2022-04-07 12:35:12       0   
236        279608  257791 2022-03-21 17:56:03 2022-03-21 18:36:19       0   
378        302158  421044 2022-04-11 13:00:17 2022-04-11 15:26:25       0   
379        273249  248722 2022-03-15 12:49:03 2022-03-15 18:40:54       0   

     initial_amount  final_amount  retailer_id     status  agent_id  \
114          1000.0           0.0       192777     FAILED      2744   
235          1000.0        1000.0        95926  COLLECTED      1550   
236          1000.0        1000.0        95926  COLLECTED      2570   
378          1500.0        1500.0        54854  COLLECTED      2891   
379          1500.0        1500.0        54854  COLLECTED      1621   

    district_id response_in_hours  
114          28          01:22:07  
235          28          13:10:08  
236          28          00:40:16  
378          28          02:26:08  
379          28          05:51:51

In [125]:
df_district_28["response_in_hours"].mean()

Timedelta('0 days 15:38:06.176123')

also the average of time response is not that high 

In [126]:
df_district_28["status"].value_counts(normalize = True)

COLLECTED              0.882855
FAILED                 0.079520
PARTIALLY_COLLECTED    0.037625
Name: status, dtype: float64

the percentage of the failed is not big comparing the the total 

In [127]:
# find out numbers of retailers that located in this district
df_district_28["retailer_id"].nunique()

349

we will get the district data where status = failed to find out number or retailers that the agents failed to complete the order in with in this district

In [128]:
df_28_failed = df_district_28.query("status == 'FAILED'")
df_28_failed.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
114         283485  264604 2022-03-24 18:57:04 2022-03-24 20:19:11       0   
382         266836  231861 2022-03-07 19:24:03 2022-03-08 13:51:11       1   
397         218129   64606 2021-12-23 19:48:03 2021-12-25 16:06:08       0   
1303        204297   11658 2021-11-27 11:44:02 2021-11-27 16:10:49       0   
1306        302406  412436 2022-04-10 13:51:03 2022-04-11 20:57:21       1   

      initial_amount  final_amount  retailer_id  status  agent_id district_id  \
114           1000.0           0.0       192777  FAILED      2744          28   
382           1500.0           0.0        54854  FAILED      1621          28   
397           2000.0           0.0       198700  FAILED      1621          28   
1303          1000.0           0.0        64413  FAILED      1656          28   
1306          1500.0           0.0        64413  FAILED      2907          28   

     response_in_hours  
114    0 days 01:22:07  
382    0 days 18:27:08  
397    1 days 20:18:05  
1303   0 days 04:26:47  
1306   1 days 07:06:18

In [129]:
# we will find out the number of retailers 
df_28_failed["retailer_id"].nunique()

177

from the previous analysis it showed that 50.7 % of the total retailers in the district 28 has at least one failed status. so many fails happen with many retailers at least once in this district

------------------------------------------------------------------------------------------------------------------------------

we will make analysis regarding the retailer id number 48353 as it showed that 31 failed order happened with this specific retailer

In [130]:
# get only the data where retailer_id = 48353
df_retailer_48353 = df.query("retailer_id == 48353")

In [131]:
df_retailer_48353.head()

run_sheet_id   co_id          created_at          updated_at  delays  \
127669        298657  390182 2022-04-07 21:02:16 2022-04-08 13:40:57       1   
127670        294028  322200 2022-04-02 22:26:06 2022-04-04 02:38:38       1   
127671        291668  296872 2022-03-31 18:34:13 2022-04-01 14:27:20       1   
127672        288677  274723 2022-03-28 17:45:04 2022-03-29 16:44:43       1   
127673        289938  278843 2022-03-29 18:05:11 2022-03-30 18:20:05       1   

        initial_amount  final_amount  retailer_id  status  agent_id  \
127669          824.03           0.0        48353  FAILED      2799   
127670          825.00           0.0        48353  FAILED      2860   
127671          824.03           0.0        48353  FAILED      2577   
127672          825.00           0.0        48353  FAILED      2577   
127673          824.03           0.0        48353  FAILED      2326   

       district_id response_in_hours  
127669          69   0 days 16:38:41  
127670          69   1 days 04:12:32  
127671          69   0 days 19:53:07  
127672          69   0 days 22:59:39  
127673          69   1 days 00:14:54

In [132]:
df_retailer_48353["status"].value_counts()

FAILED                 31
PARTIALLY_COLLECTED     4
COLLECTED               3
Name: status, dtype: int64

as we can see the failed status is too large comparing to the total orders that occurred with this specific retailer

In [133]:
# we fill find out in which district he located in 
df_retailer_48353["district_id"].unique()

array([69.0], dtype=object)

he is located in the distrit id number 69, we will make further analysis about this district to see if the issue is with the district in general or its just the retailer 

In [134]:
# get only the data where district_id = 69
df_district_69 = df.query("district_id == 69")
df_district_69.head()


run_sheet_id   co_id          created_at          updated_at  delays  \
714         303470  432252 2022-04-12 15:07:03 2022-04-12 15:10:16       0   
2956        292902  306608 2022-04-01 15:22:10 2022-04-02 17:18:38       0   
2957        267876  234675 2022-03-08 13:08:03 2022-03-09 17:52:11       0   
2958        280558  257402 2022-03-21 14:47:03 2022-03-22 21:24:08       0   
2959        264929  228064 2022-03-06 15:35:03 2022-03-06 18:14:57       0   

      initial_amount  final_amount  retailer_id               status  \
714           1000.0         700.0        20790  PARTIALLY_COLLECTED   
2956          1500.0        1500.0        51297            COLLECTED   
2957          1500.0        1500.0        51297            COLLECTED   
2958          1500.0        1500.0        51297            COLLECTED   
2959          1500.0        2000.0        51297            COLLECTED   

      agent_id district_id response_in_hours  
714        428          69   0 days 00:03:13  
2956      2326          69   1 days 01:56:28  
2957      2250          69   1 days 04:44:08  
2958      2681          69   1 days 06:37:05  
2959      2328          69   0 days 02:39:54

In [135]:
# we will find how many retailers located in this district 
df_district_69["agent_id"].nunique()

88

In [136]:
# we will check the status values for district 69
df_district_69["status"].value_counts()

COLLECTED              1819
PARTIALLY_COLLECTED     237
FAILED                  204
Name: status, dtype: int64

so from here we can say that this district has 88 retailers and 15 % of the failed orders in this district happened with only the retailer 48353

In [137]:
# get the data with only success status where district id =69
df_success_69 = df_success.query("district_id == 69")

In [138]:
# getting number of succeeded orders in district 69
df_success_69.shape

(1819, 12)

In [146]:
# getting number of total orders in district 69
df_district_69.shape

(2260, 12)

for here we can say 80.4 % of the total orders in district 69 ended with a success 

---------------------------------------------------------------------------------------------------------------------------

there is something seems little odd and needs further analysis, which is the difference bewteen updated time and created time at some district is almost 94 days, which doesnt make sense actually, so we will make more analysis about this ditrict

In [141]:
# finding out which district that has time response > 90 days 
df[df["response_in_hours"] >"90 days"]

run_sheet_id  co_id          created_at          updated_at  delays  \
14171        247174   6862 2021-11-10 13:34:03 2022-02-12 15:42:21       0   
14172        247174   6940 2021-11-10 17:14:11 2022-02-12 15:42:13       0   
14173        247174   6932 2021-11-10 16:55:03 2022-02-12 15:42:17       0   
14174        247174   6960 2021-11-10 18:15:03 2022-02-12 15:42:08       0   

       initial_amount  final_amount  retailer_id  status  agent_id  \
14171          1000.0           0.0       161115  FAILED       493   
14172          4000.0           0.0       161115  FAILED       493   
14173          1000.0           0.0       161115  FAILED       493   
14174          1000.0           0.0       161115  FAILED       493   

      district_id response_in_hours  
14171          86  94 days 02:08:18  
14172          86  93 days 22:28:02  
14173          86  93 days 22:47:14  
14174          86  93 days 21:27:05

of course all the status in this case is a fail

In [142]:
# make analysis about the district_id number 86
df_district_86 = df.query("district_id == 86")

In [143]:
# checking the new data shape
df_district_86.shape

(1344, 12)

In [144]:
# exploring status column
df_district_86["status"].value_counts()

COLLECTED              1108
FAILED                  128
PARTIALLY_COLLECTED     108
Name: status, dtype: int64

the numbers here seem reasonable

In [145]:
df_district_86["response_in_hours"].mean()

Timedelta('0 days 23:04:45.198660')

the average time in response is high because of these 4 rows, and it could be just outliers